### ISSUE TITLE
[DX] Logging deprecations

### ISSUE THREAD

**Issue by nicoschoenmaker** [symfony/symfony#17858]
TL;DR; There should be an easy way to see whether deprecation errors occur in production, so you can know whether you're really ready for Symfony 3.

---

We are busy updating our applications to Symfony 3. The simple approach of this problem would be to
* Fix the most obvious deprecations that turn up in the dev bar.
* Update to Symfony 3.
* Test.

However, some of our applications are rather big. What we would like to do is
* First fix the most obvious deprecations that turn up in the dev bar
* Turn on the logging of deprecation errors to the server logs.
* Read deprecation errors from the logs & fix them.
* Update to Symfony 3.
* Test.

For the first step the developer experience is rather good, so thanks for that! The 3th and 4th step should also be no problem. My problem is with the second step, logging them.

When the initial deprecations were added, all deprecations were logged  and people [ran out of diskspace](https://github.com/symfony/symfony/pull/14900). The logging should clearly not be on by default. But right now they have been turned off rather rigorously. How can we still log them to the server log?

### Solution 1: By adding a user land error handler
The first thing I tried was to add a user land error handler. It's rather easy to do this, see [this gist](https://gist.github.com/nicoschoenmaker/02386b968893adcd465e).

When I do this, suddenly *all* deprecation logs dissapear from the toolbar! After a bit of tracing I found that the DebugHandlersListener only injects the logger if the current handler is an error handler. Relevant links to source: [injecting logger](https://github.com/symfony/symfony/blob/master/src/Symfony/Component/HttpKernel/EventListener/DebugHandlersListener.php#L73-L92), [injecting exception handler](https://github.com/symfony/symfony/blob/master/src/Symfony/Component/HttpKernel/EventListener/DebugHandlersListener.php#L113-L117), [also injecting exception handler](https://github.com/symfony/symfony/blob/master/src/Symfony/Component/Debug/ExceptionHandler.php#L59-L62).

Solution one would mean creating a bundle that mimicks the behaviour of the DebugHandlersListener to inject the logger into my error handler, so it can forward it to the one from Symfony. It would also mean that everyone updating a large app has to reinvent this wheel.

### Solution 2: Solving it in Symfony
What if we'd add an option to the framework bundle (or another?) `log_deprecations`.

```
framework:
    log_deprecations: true # default false
```

Which could then do something similar to [this gist](https://gist.github.com/nicoschoenmaker/02386b968893adcd465e). This would give me the level of DX that I'd expect from Symfony.

### Solution 3: Reinstating the ErrorsLoggerListener
The (now deprecated) ErrorsLoggerListener [injected](https://github.com/symfony/symfony/blob/2.8/src/Symfony/Component/HttpKernel/EventListener/ErrorsLoggerListener.php#L42-L45) the logger in a way that makes it possible to replace the error handler.

### Solution 4: The one I've missed?
* Searching for everything in the 30 page [upgrade guide](https://github.com/symfony/symfony/blob/master/UPGRADE-3.0.md) isn't really an option for us.
* Our code coverage is increasing, so the PHPUnit deprecations help. But that alone doesn't give us enough confidence to upgrade.
* The [upgrade guide](http://symfony.com/doc/current/cookbook/upgrade/major_version.html#make-your-code-deprecation-free) only offers the simple approach here.

What is the best way to log the deprecation notices server side? If something already exists, that would be awesome! If someone proposes a possible Symfony side solution, I'd be willing to work towards a PR for the 2.8 (or 2.7?) branch.

**Comment by jameshalsall** [comment]
Given that we cannot add new features to 2.x (only master) this won't offer much value if it's offered by Symfony until upgrading to 4.0.

It might be worth implementing this as a standalone bundle and registering it as a log handler in monolog.

**Comment by xabbuh** [comment]
This is now possible with #19568, right? ping @lyrixx @nicolas-grekas

**Comment by lyrixx** [comment]
Yes. All deprecation are now logged by default. But it's in Symfony 3.2+ ...

**Comment by nicoschoenmaker** [comment]
We've migrated to Symfony 3 by now, so (for me) this issue can be closed if migrating to Symfony 4 is possible in the way I described above.
